### PR TITLE
RFC: Support ReverseDiff

### PR DESCRIPTION
**Pull request by anriseth**
This adds ReverseDiff automatic differentiation support.

I am not sure whether I have set it up in the most robust way, some feedback on how I call ReverseDiff will be appreciated ( @jrevels ?) 

I have chosen to change `autodiff::Bool` to be a symbol, and allow the options `autodiff == :finite`, `:forward` or `:reverse`. Anybody opposed to this approach?

TODO:
- [ ] Deprecate `autodiff::Bool` keywords

 @jeff-regier

### CODE REVIEW

File: src/optimize.jl [hunk 1]
```diff
@@ -111,33 +111,47 @@ function optimize{F<:Function, T, M <: Union{FirstOrderSolver, SecondOrderSolver
             h! = (x, out) -> ForwardDiff.hessian!(out, f, x, hcfg)
             d = TwiceDifferentiable(f, g!, fg!, h!)
         end
-    end
+    elseif options.autodiff == :reverse
```
Comment (anriseth) [inline src/optimize.jl]: @jrevels Is this way of implementing  `g!` and `fg!` and `h!` sensible?

**Comment by mlubin** [comment]
``autodiff=true`` is dead. Long live ``autodiff=true``

File: src/optimize.jl (hunk 1 above)
Comment (anriseth) [inline src/optimize.jl]: (I mean in the lines 114-132)

**Comment by jrevels** [comment]
First off, thanks for this!

Getting into the nitty gritty: Calling `ReverseDiff.compile_*` in non-top-level scope is generally a bad idea.  I've been meaning to emphasize this in the documentation more, but haven't gotten around to it - sorry about that.  On Julia v0.6, this actually won't even be usable (or at least shouldn't be usable) because of ReverseDiff compilation depends on `eval` (since it essentially translates runtime values into a compiled function).

Instead, you can use the `ReverseDiff.gradient!(result, input, cfg)` style API methods (making sure, of course, to preallocate reuse `cfg` where possible). 

To provide a better explanation: You're guaranteeing certain properties about a function when you compile it using ReverseDiff. These properties are in the ReverseDiff docs, but might be a bit scattered (again, my bad). For example, one of the guarantees is that the function won't generate a tape that is too long to be compiled in a reasonable amount of time. Another is that the function doesn't have any value-dependent branches. If you're using `ReverseDiff.compile_*` in a non-top-level scope, it's probably because you don't actually know what your objective function is, and thus, you can't guarantee that it's safe to compile.

**Comment by anriseth** [comment]
>Getting into the nitty gritty: Calling ReverseDiff.compile_* in non-top-level scope is generally a bad idea. I've been meaning to emphasize this in the documentation more, but haven't gotten around to it - sorry about that. On Julia v0.6, this actually won't even be usable (or at least shouldn't be usable) because of ReverseDiff compilation depends on eval (since it essentially translates runtime values into a compiled function).

Ah, yes I was wondering if `compile_` was a bit too optimistic. 

> Instead, you can use the ReverseDiff.gradient!(result, input, cfg) style API methods (making sure, of course, to preallocate reuse cfg where possible).

Is it safe to preallocate cfg for ReverseDiff, in the same way that Optim does forForwardDiff? https://github.com/JuliaOpt/Optim.jl/blob/master/src/optimize.jl#L98

**Comment by anriseth** [comment]
In any case, this PR is not necessarily the way we want to go with AD, based on the most recent discussions in #329.

**Comment by pkofod** [comment]
> In any case, this PR is not necessarily the way we want to go with AD, based on the most recent discussions in #329.

Well, we might still want to provide "wrappers" as @jrevels mentioned in #329 , and this code is very relevant there! So your work will definitely not have been for nothing.

**Comment by pkofod** [comment]
thanks, closed by #384

**Pull request closed without merging**